artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingManagerImpl.java

```diff
@@ -583,22 +591,65 @@ public void forEachTransaction(BiConsumer<Long, PageTransactionInfo> transaction
    public Future<Object> rebuildCounters(Set<Long> storedLargeMessages) {
       Map<Long, PageTransactionInfo> transactionsSet = new LongObjectHashMap();
       // making a copy
-      transactions.forEach(transactionsSet::put);
+      transactions.forEach((a, b) -> {
+         transactionsSet.put(a, b);
+         b.setOrphaned(true);
+      });
+      AtomicLong minLargeMessageID = new AtomicLong(Long.MAX_VALUE);
+
+      // make a copy of the stores
+      Map<SimpleString, PagingStore> currentStoreMap = new HashMap<>();
+      stores.forEach(currentStoreMap::put);
 
       if (logger.isDebugEnabled()) {
          logger.debug("Page Transactions during rebuildCounters:");
          transactionsSet.forEach((a, b) -> logger.debug("{} = {}", a, b));
       }
 
-      stores.forEach((address, pgStore) -> {
-         PageCounterRebuildManager rebuildManager = new PageCounterRebuildManager(this, pgStore, transactionsSet, storedLargeMessages);
+      currentStoreMap.forEach((address, pgStore) -> {
+         PageCounterRebuildManager rebuildManager = new PageCounterRebuildManager(this, pgStore, transactionsSet, storedLargeMessages, minLargeMessageID);
          logger.debug("Setting destination {} to rebuild counters", address);
          managerExecutor.execute(rebuildManager);
       });
 
+      managerExecutor.execute(() -> cleanupPageTransactions(transactionsSet, currentStoreMap));
+
       FutureTask<Object> task = new FutureTask<>(() -> null);
       managerExecutor.execute(task);
 
       return task;
    }
+
+   private void cleanupPageTransactions(Map<Long, PageTransactionInfo> transactionSet, Map<SimpleString, PagingStore> currentStoreMap) {
+      if (server == null) {
+         logger.warn("Server attribute was not set, cannot proceed with page transaction cleanup");
+      }
+      AtomicBoolean proceed = new AtomicBoolean(true);
+      currentStoreMap.forEach((a, b) -> {
+         if (!b.getCursorProvider().isRebuildDone()) {
+            logger.warn("cannot proceed on cleaning up page transactions as page cursor for {} is not done rebuilding it", b.getAddress());
+            proceed.set(false);
+         }
+      });
+
+      if (!proceed.get()) {
```

Comment:
Is there some additional handling to retry or is this cleanup left for another restart?  It's unclear to me what the state is after the code determines it can't proceed.

Reply:
In case one of the PageStore cleanup failed for any reason, I would prefer to never do this cleanup, and only try it at a next restart.

It shouldn't really happen, but I wanted to add this check here.

Reply:
Might be worth a comment explaining the thinking as I starred at it for awhile trying to figure out where it was being restarted etc.

Reply:
I just did. thanks